x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/ApiKeyServiceTests.java

```diff
@@ -2029,6 +2030,126 @@ public void testGetApiKeyMetadata() throws IOException {
+        final TransportVersion minNodeVersion = randomFrom(
+            Version.getDeclaredVersions(Version.class)
+                .stream()
+                .filter(v -> v.transportVersion.onOrAfter(RoleDescriptor.VERSION_REMOTE_INDICES))
+                .map(v -> v.transportVersion)
+                .toList()
+        );
+
+        final Set<RoleDescriptor> result = ApiKeyService.maybeRemoveRemoteIndicesPrivilegesFromUserRoleDescriptors(
+            userRoleDescriptors,
+            requestedRoleDescriptors,
+            minNodeVersion,
+            apiKeyId
+        );
+
+        // User roles should be unchanged.
+        assertThat(result, equalTo(userRoleDescriptors));
+    }
+
+    private static RoleDescriptor randomRoleDescriptorWithRemoteIndexPrivileges() {
+        return new RoleDescriptor(
+            randomAlphaOfLengthBetween(3, 90),
+            randomSubsetOf(ClusterPrivilegeResolver.names()).toArray(String[]::new),
+            RoleDescriptorTests.randomIndicesPrivileges(0, 3),
+            RoleDescriptorTests.randomApplicationPrivileges(),
+            RoleDescriptorTests.randomClusterPrivileges(),
+            generateRandomStringArray(5, randomIntBetween(2, 8), false, true),
+            RoleDescriptorTests.randomRoleDescriptorMetadata(randomBoolean()),
+            Map.of(),
+            RoleDescriptorTests.randomRemoteIndicesPrivileges(1, 3)
+        );
+    }
+
+    public void testMaybeRemoveRemoteIndicesPrivilegesWithRequestedRoleDescriptorsWithoutRemoteIndices() {
+        final String apiKeyId = randomAlphaOfLengthBetween(5, 8);
+        final Set<RoleDescriptor> userRoleDescriptors = Set.copyOf(
+            randomList(1, 3, () -> RoleDescriptorTests.randomRoleDescriptor(randomBoolean(), randomBoolean()))
+        );
+
+        // Requested role descriptors do not include remote indices privileges,
+        // hence we should always remove remote indices privileges (if any) from user's role descriptors.
+        final List<RoleDescriptor> requestedRoleDescriptors = randomList(
+            1,
+            2,
+            () -> RoleDescriptorTests.randomRoleDescriptor(randomBoolean(), false)
+        );
+
+        // Selecting random supported version.
+        final TransportVersion minNodeVersion = randomFrom(
+            Version.getDeclaredVersions(Version.class)
+                .stream()
+                .filter(v -> v.transportVersion.onOrAfter(RoleDescriptor.VERSION_REMOTE_INDICES))
+                .map(v -> v.transportVersion)
+                .toList()
+        );
+
+        final Set<RoleDescriptor> result = ApiKeyService.maybeRemoveRemoteIndicesPrivilegesFromUserRoleDescriptors(
+            userRoleDescriptors,
+            requestedRoleDescriptors,
+            minNodeVersion,
+            apiKeyId
+        );
+        assertThat(result.stream().anyMatch(RoleDescriptor::hasRemoteIndicesPrivileges), equalTo(false));
+        assertThat(result.size(), equalTo(userRoleDescriptors.size()));
+    }
+
```

Comment:
It would be great to cover some of the above cases more end-to-end, ideally in an IT (would have to figure out how to force a low min version, I'm not sure off the top of my head how), or here for createApiKey and updateApiKey.

Reply:
I'm planning to write more IT tests, but would prefer to do it in a follow up PR where I'll have full end-to-end authentication working (https://github.com/elastic/elasticsearch/pull/93808).  Let me know what you think!

Reply:
I've added a [BWC IT test](https://github.com/elastic/elasticsearch/blob/4fe0aa0ca19d23417dcc8e7ff6538d2b064f53aa/x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/ApiKeyBackwardsCompatibilityIT.java) for creating and updating API keys during rolling upgrade. I'm planning to add more for mixed CCS in a followup PR.